### DISCUSSION TITLE
Supporting domain-specific forks

### DISCUSSION THREAD

**Discussion by lalligagger** [TheR1D/shell_gpt#110] [Ideas]
Hi, very cool project! It's definitely given me inspiration for a project I'm calling `satgpt` which would be a domain-specific helper for satellite modeling and image retrieval.

I'd like to use `sgpt` as the core interface but will need to adapt/ overwrite the prompts. I was also thinking to point to a specific list of "allowed" or at least priority packages to help guide responses. Lastly, I have the need to add some docs for some of my own CLI tools, etc.

My question for now is if anyone has thought of a good way to do this, that is more elegant than just forking into my project. (Where I expect it'd get outdated quickly, but I would only need to change the app and prompt generator, so plan to just manage that manually for now.)

Anyway, thanks for a great project and I'm looking forward to exploring some more!

**Comment by lalligagger** [comment]
Figured I'd add that my only reason for changing app.py would be to still prepend context for plain text responses. So maybe that could be an empty string in this repo, and the user could override any of `SHELL` `CODE` or `ASSISTANT` pre-prompts?

**Reply by lpurdy01** [reply]
In the conversation chain for my pull request #92 I have proposed a system for custom roles to be set and their management. I think with some kind of save file structure this could be extended to do what you want. 

I'm interested in ideas on how it should be implemented, and exploring tangential use cases will help inform a lot.

**Reply by lalligagger** [reply]
That PR is definitely in the right direction, thanks! I will test out the custom roles once it's merged.

**Reply by lalligagger** [reply]
@lpurdy01 saw your PR is taking some time, anything I can test out locally ahead of releasing my package?

**Comment by lalligagger** [comment]
So far, so good running some quick experiments with the new role features!

<img width="1434" alt="image" src="https://user-images.githubusercontent.com/11700267/232255580-54fea90c-83ab-44d2-9fc1-e031166b11ae.png">

**Reply by TheR1D** [reply]
You can also use `--shell` to execute shell commands right away. For better suggestions you can specify something like **"stac-client command"** as role output type instead of just **"Command"**

**Reply by lalligagger** [reply]
Thanks, this was just a quick test to A/B the roles. Really liking the new repl! Works great for this use case with tends to have lots of steps in sequence.

Done for today now that I have `satgpt` building locally, and expect it will publish just fine once the role updates are released here.
